lib/Backends/OpenCL/OpenCL.cpp

```diff
@@ -1137,36 +1160,77 @@ Error OpenCLFunction::execute(ExecutionContext *context) {
-      setKernelArg(kernel, numArgs + 3, pads);
-      setKernelArg(kernel, numArgs + 4, CC->getGroup());
-      setKernelArg(kernel, numArgs + 5, CC->getDilation());
-      setKernelArg(kernel, numArgs + 6, odim);
-      setKernelArg(kernel, numArgs + 7, idim);
-      setKernelArg(kernel, numArgs + 8,
-                   ShapeNHWC(CC->getFilter()->getType()->dims()));
-      if (isQuantized) {
-        auto srcTy = CC->getSrc()->getType();
-        auto destTy = CC->getDest()->getType();
-        auto filterTy = CC->getFilter()->getType();
-        auto biasTy = CC->getBias()->getType();
-        setKernelArg(kernel, numArgs + 9, destTy->getOffset());
-        setKernelArg(kernel, numArgs + 10, destTy->getScale());
-        setKernelArg(kernel, numArgs + 11, srcTy->getOffset());
-        setKernelArg(kernel, numArgs + 12, srcTy->getScale());
-        setKernelArg(kernel, numArgs + 13, filterTy->getOffset());
-        setKernelArg(kernel, numArgs + 14, filterTy->getScale());
-        setKernelArg(kernel, numArgs + 15, biasTy->getOffset());
-        setKernelArg(kernel, numArgs + 16, biasTy->getScale());
+      auto odim = ShapeNHWC(CC->getDest()->getType()->dims());
+      ShapeNHWC kernelSize(CC->getFilter()->getType()->dims());
+      auto pads = PaddingTLBR(CC->getPads());
+
+      std::vector<std::string> options;
+
+      const bool specialize = clSpecializeConvolution && !isQuantized;
+      std::string src;
+      if (specialize) {
+        // Specialize the kernel related to Conv node parameters to enable
+        // aggressive constant propagation and other optimizations.
+        addIntOption(options, "CONVK_GROUP", CC->getGroup());
+        addIntOption(options, "CONVK_BATCHES", idim.n);
+        addIntOption(options, "CONVK_DILATION", CC->getDilation());
+        addIntOption(options, "CONVK_KERNEL_W", kdim.width);
+        addIntOption(options, "CONVK_KERNEL_H", kdim.height);
+        addIntOption(options, "CONVK_STRIDES_W", sdim.width);
+        addIntOption(options, "CONVK_STRIDES_H", sdim.height);
+        addIntOption(options, "CONVK_IDIM_W", idim.w);
+        addIntOption(options, "CONVK_IDIM_H", idim.h);
+        addIntOption(options, "CONVK_IDIM_C", idim.c);
+        addIntOption(options, "CONVK_ODIM_W", odim.w);
+        addIntOption(options, "CONVK_ODIM_H", odim.h);
+        addIntOption(options, "CONVK_ODIM_C", odim.c);
+        addIntOption(options, "CONVK_PADS_TOP", pads.top);
+        addIntOption(options, "CONVK_PADS_LEFT", pads.left);
+        addIntOption(options, "CONVK_FILTER_W", kernelSize.w);
+        addIntOption(options, "CONVK_FILTER_H", kernelSize.h);
+        addIntOption(options, "CONVK_FILTER_C", kernelSize.c);
+        src.append(reinterpret_cast<const char *>(
+                       specialized_no_local_mem_conv_cl_src),
+                   specialized_no_local_mem_conv_cl_src_size);
+      } else {
+        src.append(reinterpret_cast<const char *>(kernels_cl_src),
```

Comment:
Why do we need create the source and then createProgram for the standard, non-specialized case? It seems like a waste of compute. Can we leave this execution path untouched and only pay the runtime cost for the specialized case?